### PR TITLE
[ALLUXIO-2559] Check that ramfs/tmpfs is big enough

### PR DESCRIPTION
**Pull request by aaudiber**
https://alluxio.atlassian.net/browse/ALLUXIO-2559

This PR adds best-effort checking for the case where the ramfs/tmpfs is smaller than the memory it is configured as. If the size of the ramfs/tmpfs can't be determined, no checking is done.

Example output:
```
[ec2-user@ip-172-31-74-144 alluxio]$ cat logs/worker.out
Exception in thread "main" java.lang.RuntimeException: java.lang.IllegalStateException: tmpfs is smaller than the configured size: tmpfs size: 512MB, configured size: 3GB
	at alluxio.worker.AlluxioWorkerProcess.<init>(AlluxioWorkerProcess.java:147)
	at alluxio.worker.WorkerProcess$Factory.create(WorkerProcess.java:35)
	at alluxio.worker.AlluxioWorker.main(AlluxioWorker.java:54)
Caused by: java.lang.IllegalStateException: tmpfs is smaller than the configured size: tmpfs size: 512MB, configured size: 3GB
	at alluxio.worker.block.meta.StorageTier.checkEnoughMemSpace(StorageTier.java:144)
	at alluxio.worker.block.meta.StorageTier.initStorageTier(StorageTier.java:102)
```

I recommend reading the first 3 commits individually.

### CODE REVIEW

File: core/common/src/main/java/alluxio/security/authorization/ModeParser.java [hunk 1]
```diff
@@ -120,6 +121,7 @@ private Mode parseSymbolic(String value) {
             break;
           default:
             // Should never get here as already checked for invalid targets
```
Comment (jsimsa) [inline core/common/src/main/java/alluxio/security/authorization/ModeParser.java]: nit: it bugs me that the comment and exception are in different order between the two default cases :-)

Comment (aaudiber) [inline core/common/src/main/java/alluxio/security/authorization/ModeParser.java]: good call, fixed

File: core/server/worker/src/main/java/alluxio/worker/block/meta/StorageTier.java [hunk 2]
```diff
@@ -94,6 +98,55 @@ private void initStorageTier()
...
+          continue;
+        }
+      } catch (InvalidPathException e) {
+        continue;
+      }
+
+      if ((fsType.equalsIgnoreCase("tmpfs") || fsType.equalsIgnoreCase("ramfs"))
+          && size < storageDir.getCapacityBytes()) {
+        throw new IllegalStateException(String.format(
+            "tmpfs is smaller than the configured size: tmpfs size: %s, configured size: %s",
```
Comment (jsimsa) [inline core/server/worker/src/main/java/alluxio/worker/block/meta/StorageTier.java]: message mentions `tmpfs` but if condition also allows `ramfs`

File: core/common/src/main/java/alluxio/util/CommonUtils.java [hunk 3]
```diff
@@ -426,18 +431,34 @@ public void remove() {
...
    * Executes the given callables, waiting for them to complete (or time out).
-
+   *
    * @param callables the callables to execute
    * @param <T> the return type of the callables
+   * @throws TimeoutException if one of the callables times out
+   * @throws Exception if one of the callables throws an exception
    */
-  public static <T> void invokeAll(List<Callable<T>> callables) {
+  public static <T> void invokeAll(List<Callable<T>> callables) throws TimeoutException, Exception {
```
Comment (calvinjia) [inline core/common/src/main/java/alluxio/util/CommonUtils.java]: In the case one of the callables throws an exception, there is no guarantee that all the callables have run before returning?

File: core/common/src/main/java/alluxio/util/ShellUtils.java [hunk 4]
```diff
@@ -56,108 +65,127 @@
...
+      builder.setFsType(match.replace("type", "").trim());
+      lineWithoutType = line.replace(match, "");
+    } else {
+      lineWithoutType = line;
+    }
+    // Now parse the rest
+    matcher = Pattern.compile("(.*) on (.*) \\((.*)\\)").matcher(lineWithoutType);
+    if (!matcher.matches()) {
+      LOG.debug("Unable to parse output of 'mount': {}", line);
+      return builder.build();
```
Comment (calvinjia) [inline core/common/src/main/java/alluxio/util/ShellUtils.java]: Is the returned `UnixMountInfo` still valid?

Comment (aaudiber) [inline core/common/src/main/java/alluxio/util/ShellUtils.java]: Yes, all fields are optional. I've updated to use Optional to reflect this.

File: core/common/src/main/java/alluxio/util/CommonUtils.java (hunk 3 above)
Comment (aaudiber) [inline core/common/src/main/java/alluxio/util/CommonUtils.java]: I made some updates to the method and added unit tests. The guarantee now is that when the method returns, either every callable has completed, or we have made a best effort to halt it by calling `shutdownNow()`, which will usually interrupt the callable.

File: core/server/worker/src/main/java/alluxio/worker/block/meta/StorageTier.java (hunk 2 above)
Comment (aaudiber) [inline core/server/worker/src/main/java/alluxio/worker/block/meta/StorageTier.java]: good catch, updated

**Comment by aaudiber** [comment]
@calvinjia All comments addressed, PTAL

File: core/server/worker/src/main/java/alluxio/worker/block/meta/StorageTier.java [hunk 5]
```diff
@@ -94,6 +99,58 @@ private void initStorageTier()
...
+    } catch (IOException e) {
+      LOG.debug("Failed to get mount information: {}", e.getMessage());
+      return;
+    }
+    for (UnixMountInfo mountInfo : info) {
+      Optional<String> mountPointOption = mountInfo.getMountPoint();
+      Optional<String> fsTypeOption = mountInfo.getFsType();
+      Optional<Long> sizeOption = mountInfo.getOptions().getSize();
+      if (!mountPointOption.isPresent() || !fsTypeOption.isPresent() || !sizeOption.isPresent()) {
+        continue;
```
Comment (calvinjia) [inline core/server/worker/src/main/java/alluxio/worker/block/meta/StorageTier.java]: Should we log something here to inform the user that it is not being validated?

Comment (aaudiber) [inline core/server/worker/src/main/java/alluxio/worker/block/meta/StorageTier.java]: done

**Pull request merged**